metapool/prep.py

```diff
@@ -938,3 +907,178 @@ def demux_pre_prep(pre_prep):
+        # should be of the form:
+        # <project_name_w_qiita_id>/filtered_sequences/<fastq_file>
+        tmp = fastq_fp.replace(fp, '')
+        # remove any leading and/or trailing '/' characters from the
+        # remaining path.
+        # use sep instead of '/' to be more platform independent.
+        tmp = tmp.strip(sep)
+        tmp = tmp.split(sep)
+
+        # tmp[1] != 'filtered_sequences' doesn't appear to be possible given
+        # the glob statement above, but we'll keep it here for safety.
+        if len(tmp) != 3 or tmp[1] != 'filtered_sequences':
+            raise ValueError(f"{fastq_fp} appears to be stored in an "
+                             "unexpected location")
+
+        # where tmp[0] is the project_name, keep a list of all associated with
+        # that project.
+        by_project[tmp[0]].append(full_path)
+        run_prefix_mapping[full_path] = run_prefix
+        count += 1
+
+    if count == 0:
+        raise ValueError(f"No fastq.gz files were found in {fp}. Please "
+                         "ensure that the path is correct and that the "
+                         "directory contains fastq.gz files.")
+
+    # convert to standard dict before returning.
+    return dict(by_project), run_prefix_mapping
+
+
+def get_run_prefix(file_name):
+    # Modified from mg-scripts/util.py:determine_orientation().
+    orientations = ['R1', 'R2', 'I1', 'I2']
+
+    results = []
+
+    # assume orientation is always present in the file's name.
+    # assume that it is of one of the four forms above.
+    # assume that it is always the right-most occurance of the four
+    # orientations above.
+    # assume that orientation is encapsulated with either '_' or '.'
+    # e.g.: '_R1_', '.I2.'.
+    # assume users can and will include any or all of the four
+    # orientation as part of their filenames as well. e.g.:
+    # ABC_7_04_1776_R1_SRE_S3_L007_R2_001.trimmed.fastq.gz
+    for o in orientations:
+        variations = [f"_{o}_", f".{o}."]
+        for v in variations:
+            # rfind searches from the end of the string, rather than
+            # its beginning. It returns the position in the string
+            # where the substring begins.
+            results.append((file_name.rfind(v), o))
+
+    # the orientation will be the substring found with the maximum
+    # found value for pos. That is, it will be the substring that
+    # begins at the rightest most position in the file name.
+    results.sort(reverse=True)
+
+    pos, orientation = results[0]
+
+    # if no orientations were found, raise an Error.
+    if pos == -1:
+        raise ValueError(f"Could not determine orientation of {file_name}")
+
+    return file_name[0:pos], orientation
+
+
+def is_nonempty_gz_file(name):
+    """Taken from https://stackoverflow.com/a/37878550/379593"""
+    with gzip.open(name, 'rb') as f:
+        try:
+            file_content = f.read(1)
+            return len(file_content) > 0
+        except Exception:
```

Comment:
Should we check for fastq.gz files or just gz files?

Reply:
I removed this function originally and restored it in this location because I realized it was needed for the amplicon pathway. That's its only current user.